src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/LengthAttribute.cs

```diff
@@ -0,0 +1,102 @@
+{
+    /// <summary>
+    ///     Specifies the minimum and maximum length of collection/string data allowed in a property.
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Property | AttributeTargets.Field | AttributeTargets.Parameter, AllowMultiple = false)]
+    public class LengthAttribute : ValidationAttribute
+    {
+        [RequiresUnreferencedCode(CountPropertyHelper.RequiresUnreferencedCodeMessage)]
+        public LengthAttribute(int minimumLength, int maximumLength)
+            : base(SR.LengthAttribute_ValidationError)
+        {
+            MinimumLength = minimumLength;
+            MaximumLength = maximumLength;
+        }
+
+        /// <summary>
+        ///     Gets the minimum allowable length of the collection/string data.
+        /// </summary>
+        public int MinimumLength { get; }
+
+        /// <summary>
+        ///     Gets the maximum allowable length of the collection/string data.
+        /// </summary>
+        public int MaximumLength { get; }
+
+        /// <summary>
+        ///     Determines whether a specified object is valid. (Overrides <see cref="ValidationAttribute.IsValid(object)" />)
+        /// </summary>
+        /// <remarks>
+        ///     This method returns <c>true</c> if the <paramref name="value" /> is null.
+        ///     It is assumed the <see cref="RequiredAttribute" /> is used if the value may not be null.
+        /// </remarks>
+        /// <param name="value">The object to validate.</param>
+        /// <returns>
+        ///     <c>true</c> if the value is null or its length is between the specified minimum length and maximum length, otherwise
+        ///     <c>false</c>
+        /// </returns>
+        /// <exception cref="InvalidOperationException">
+        ///     <see cref="MinimumLength"/> is less than zero or <see cref="MaximumLength"/> is less than <see cref="MinimumLength"/>.
+        /// </exception>
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode", Justification = "The ctor is marked with RequiresUnreferencedCode.")]
+        public override bool IsValid(object? value)
+        {
+            // Check the lengths for legality
+            EnsureLegalLengths();
+
+            int length;
+            // Automatically pass if value is null. RequiredAttribute should be used to assert a value is not null.
+            if (value is null)
+            {
+                return true;
+            }
+            if (value is string str)
+            {
+                length = str.Length;
+            }
+            else if (CountPropertyHelper.TryGetCount(value, out int count))
+            {
+                length = count;
+            }
+            else
+            {
+                throw new InvalidCastException(SR.Format(SR.LengthAttribute_InvalidValueType, value.GetType()));
+            }
+
+            return MinimumLength <= length && length <= MaximumLength;
```

Comment:
>MinimumLength <= length && length <= MaximumLength [](http://example.com/codeflow?start=19&length=50)

This looks good. I am just wondering do we care optimizing it? I mean something like

(uint)(length - MinimumLength) <= (uint)(MaximumLength - MinimumLength)

Reply:
Could do, although generally speaking this is not a particularly performance oriented library. We might want to revisit fixing low hanging fruit more holistically, particularly if we end up adding a source generator in this layer.

Reply:
> length - MinimumLength

Note that this expression could turn out to be negative, so I don't think we could avoid doing two comparisons here.

Reply:
The negative number will be converted to `uint` which will be a very big number and still work.

Reply:
Good point. I've updated the code.